quark/core/shurikenapkinfo.py

```diff
@@ -0,0 +1,612 @@
+        descriptor: Optional[str] = None,
+    ) -> List[MethodObject]:
+        """
+        Inherited from baseapkinfo.py.
+        Find a method with the given class name, method name, and descriptor.
+
+        :param class_name: the class name of the target method. Defaults to
+        None
+        :param method_name: the method name of the target method. Defaults to
+        None
+        :param descriptor: the descriptor of the target method. Defaults to
+        None
+        :return: a list of the target MethodObject
+        """
+        methods = self.all_methods
+        if class_name:
+            methods = (m for m in methods if class_name == m.class_name)
+
+        if method_name:
+            methods = (m for m in methods if method_name == m.name)
+
+        if descriptor:
+            methods = (m for m in methods if descriptor == m.descriptor)
+
+        return list(methods)
+
+    @functools.lru_cache()
+    def upperfunc(self, method_object: MethodObject) -> Set[MethodObject]:
+        """
+        Inherited from baseapkinfo.py.
+        Find the xrefs from the specified method.
+
+        :param method_object: a target method which the returned methods
+        should call
+        :return: a set of MethodObjects
+        """
+        methodAnalysis = method_object.cache
+        upperFuncs = set()
+        for i in range(methodAnalysis.n_of_xreffrom):
+            upperFuncs.add(
+                self.__convert_to_method_object(
+                    methodAnalysis.xreffrom[i].method.contents
+                )
+            )
+
+        return upperFuncs
+
+    @functools.lru_cache()
+    def lowerfunc(
+        self, method_object: MethodObject
+    ) -> list[Tuple[MethodObject, int]]:
+        """
+        Inherited from baseapkinfo.py.
+        Find the xrefs to the specified method.
+
+        :param method_object: a target method used to find what methods it
+        calls
+        :return: a set of tuples consisting of the called method and the
+        offset of the invocation
+        """
+        methodAnalysis = method_object.cache
+
+        lowerFuncs = []
+        for i in range(methodAnalysis.n_of_xrefto):
+            xref = methodAnalysis.xrefto[i]
+            lowerFuncs.append(
+                (
+                    self.__convert_to_method_object(xref.method.contents),
+                    xref.idx,
+                )
+            )
+
+        return lowerFuncs
+
+    def get_method_bytecode(
+        self, method_object: MethodObject
+    ) -> Generator[BytecodeObject, None, None]:
+        """
+        Inherited from baseapkinfo.py.
+        Return the bytecodes of the specified method.
+
+        :param method_object: a target method to get the corresponding
+        bytecodes
+        :yield: a generator of BytecodeObjects
+        """
+        disassembledMethod = self.__getDisassembledMethod(method_object)
+        if method_object.cache.external:
+            return
```

Comment:
+ Please do the if statement before getting the disassembled method to avoid unnecessary calculations.
+ Instead of returning None, a better practice is to return a generator that generates nothing.